### PR TITLE
Switch to non-SSL if nodes are in the same zone

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Resolves [11546](https://github.com/crate/crate/issues/11546)

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by BaurzhanSakhariev** [comment]
Hi @mfussenegger, could you take a look to `switch_to_plaintext` flag and it's usage?
Probably introducing new flag is overkill and we could just use trust for intra zone communication [like we did before](https://github.com/crate/crate/commit/e1b2b87c3eeebd48359c244a8020cf5beca5e7c7#diff-068a24ef8c4490ccc059c02a8bfa68b075441f7b64c4072468f9ac3850b44e6dR213-R214)

But I added this flag after re-reading this message https://crate.slack.com/archives/CBPLN6WJK/p1626100016463800?thread_ts=1625853131.451800&cid=CBPLN6WJK
> for the cert-auth to pass, you already need a completed handshake (because it must access the client certificate

Basically with this new flag we use method: cert and go extra certificate validation otherwise we can drop this flag and just use trust

File: server/src/main/java/io/crate/auth/ClientCertAuth.java [hunk 1]
```diff
@@ -33,11 +33,22 @@
 
 public class ClientCertAuth implements AuthenticationMethod {
 
-    static final String NAME = "cert";
+    public static final String NAME = "cert";
     private final UserLookup userLookup;
 
+    // SSL is normally required for cert method,
+    // this flag represents an HBA config entry corresponding
+    // to the concrete instance of CertAuth and letting user to downgrade to non-SSL.
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/auth/ClientCertAuth.java:41]: Maybe move to the getter method as regular java docstring - so that the information is visible in editors/IDEs when they call the method

File: server/src/main/java/org/elasticsearch/transport/netty4/AdaptiveSslHandler.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws SSLException {
+        // Inspired by OptionalSslHandler.decode.
+
+        if (in.readableBytes() < SSL_RECORD_HEADER_LENGTH) {
+            return;
+        }
+
+        if (SslHandler.isEncrypted(in)) {
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/transport/netty4/AdaptiveSslHandler.java:58]: Is it expensive to run this check? If it is we could consider to replace the `AdaptiveSslHandler` with the regular handler after some point, and only allow the downgrade happening after the initial handshake

File: docs/config/node.rst [hunk 3]
```diff
@@ -485,6 +485,15 @@ The meaning of the fields of the are as follows:
...
   | information about these methods.
 
+.. _auth.host_based.config.${order}.switch_to_plaintext:
+
+**auth.host_based.config.${order}.switch_to_plaintext**
+  | *Default:* ``false``
+  | *Runtime:* ``no``
+
+  | Specifies whether node can switch to plaintext after establishing secured
+  | connection.
```
Comment (mfussenegger) [inline docs/config/node.rst:495]: Maybe we should point out that this should really only be used in combination with some IP restrictions

File: server/src/main/java/org/elasticsearch/transport/netty4/AdaptiveSslHandler.java (hunk 2 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/org/elasticsearch/transport/netty4/AdaptiveSslHandler.java:58]: It analyses only first 5 bytes - internally it calls Netty's `getEncryptedPacketLength` which is basically same with open search check ([our adaptation](https://github.com/crate/crate/commit/a7416f9861a12a0a8c8bf9ac57c2bbbf69f843da#diff-1c5d1afa88e5576595f2438652d0bc5bd960477875add1f6e0b6a06341765de1R103-R104))
Anyway, if can improve, let's go for it - I can create a follow up PR to keep this one small if it's fine

Comment (mfussenegger) [reply server/src/main/java/org/elasticsearch/transport/netty4/AdaptiveSslHandler.java:58]: Okay - sounds like it isn't worth optimizing

File: server/src/main/java/org/elasticsearch/transport/netty4/HostBasedAuthHandler.java [hunk 4]
```diff
@@ -75,6 +84,24 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
...
+                // HBA.SSL optional in combination with method cert indicates that it's safe to downgrade to plaintext.
+                SslHandler sslHandler = (SslHandler) ctx.pipeline().get(SERVER_SSL_HANDLER_NAME);
+                if (sslHandler != null) {
+                    if (LOGGER.isDebugEnabled()) {
+                        LOGGER.debug("SSL switch to plaintext enabled, node {} switching from SSL to plaintext",
+                            ((InetSocketAddress) channel.localAddress()).getHostName()
+                        );
+                    }
+                    ctx.pipeline().remove(sslHandler);
+                }
```
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/main/java/org/elasticsearch/transport/netty4/HostBasedAuthHandler.java]: Done - created a small helper method as closing block used somewhere else.

File: server/src/main/java/io/crate/protocols/ssl/SslSettings.java [hunk 5]
```diff
@@ -73,15 +72,14 @@ static SSLMode parse(String value) {
                 case "true" -> ON;
                 case "off" -> OFF;
                 case "false" -> OFF;
-                case "legacy" -> LEGACY;
```
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/protocols/ssl/SslSettings.java]: Got it, thanks. Done in a separate last commit.

**Review by mfussenegger — APPROVED** [review]
